### PR TITLE
Feature/SK-1261 | Split Learning in FEDn

### PR DESCRIPTION
**Pull request by FrankJonasmoelle**
**This is a first implementation of Split Learning (for vertical federated learning) in FEDn.**

The implementation includes:
- A new session "start_splitlearning_session", which can be executed from the API client via "start_splitlearning_session"
- A new helper (splitlearninghelper) responsible for saving & loading embeddings and gradients
- A new aggregator (splitlearningagg) that concatenates client embeddings, continues the forward pass and calculates gradients based on the labels it receives
- The logic is based on a **forward** and **backward** pass
- In the forward pass, clients generate embeddings and send them to a combiner, where they are concatenated and the gradients are calculated
- In the backward pass, the calculated gradients are sent back to the clients, who update their model parameters by continuing the backward pass
- Validation is performed using a single forward pass, without gradient calculation and model updates (can also be used for inference)
- A vertical federated learning example based on the diabetes data set for demonstration purposes

Limitations:
- The current implementation works for only one combiner
- It is assumed that the data & labels are sorted by index to match
- Training data is processed in one batch

### CODE REVIEW

**Review by Wrede — CHANGES_REQUESTED** [review]
impressive, however I wonder if you could implement the agg using server-functions instead? This because of the dep on torch, same for helper but we currently don't have a way to user-define helper except from source. Also the PR is up do date with master? Make sure the new DTOs work.

File: examples/splitlearning_diabetes/data/diabetes.csv
[the remark this answers is not in the document]
Comment (FrankJonasmoelle) [reply examples/splitlearning_diabetes/data/diabetes.csv:1]: It's 24 KB, uploaded it for convenience

File: fedn/network/api/client.py [hunk 1]
```diff
@@ -694,6 +694,89 @@ def start_session(
         _json = response.json()
         return _json
 
+    def start_splitlearning_session(
```
[the remark this answers is not in the document]
Comment (FrankJonasmoelle) [reply fedn/network/api/client.py:697]: The logical difference is that a normal session requires a seed model, whereas this is not the case for a splitlearning session. The check for an initial seed model in "session_routes.py" is removed for splitlearning sessions. Of course, one could combine it to one function if preferred.

**Review by Wrede — COMMENTED** [review]
Status on this @FrankJonasmoelle ?

File: fedn/network/combiner/aggregators/splitlearningagg.py
[the remark this answers is not in the document]
Comment (FrankJonasmoelle) [reply fedn/network/combiner/aggregators/splitlearningagg.py:1]: Server-functions also do not allow torch as a dependency. Let's discuss this

File: fedn/network/combiner/roundhandler.py [hunk 2]
```diff
@@ -210,6 +210,93 @@ def _prediction_round(self, prediction_id: str, model_id: str, clients: list):
...
+        self.server.request_forward_pass(session_id=session_id, model_id=model_id, config=config, clients=clients)
+
+        # If buffer_size is -1 (default), the round terminates when/if all clients have completed.
+        if int(config["buffer_size"]) == -1:
+            buffer_size = len(clients)
+        else:
+            buffer_size = int(config["buffer_size"])
+
+        # Wait / block until the round termination policy has been met.
+        self.update_handler.waitforit(config, buffer_size=buffer_size)
```
[the remark this answers is not in the document]
Comment (FrankJonasmoelle) [reply fedn/network/combiner/roundhandler.py:245]: Changed it and set it so that it always waits for all clients. It is enforced that all clients need to send embeddings, otherwise the algorithm does not make sense. As SL would only be used with a low number of clients (cross-silo setting), this should be ok.

File: fedn/network/grpc/fedn_pb2.py [hunk 3]
```diff
@@ -1,8 +1,8 @@
 # -*- coding: utf-8 -*-
 # Generated by the protocol buffer compiler.  DO NOT EDIT!
 # NO CHECKED-IN PROTOBUF GENCODE
-# source: network/grpc/fedn.proto
-# Protobuf Python Version: 5.28.1
+# source: fedn.proto
+# Protobuf Python Version: 5.27.2
```
[the remark this answers is not in the document]
Comment (FrankJonasmoelle) [reply fedn/network/grpc/fedn_pb2.py]: Changed it to the exact protobuf version (5.28.1)

File: fedn/utils/helpers/plugins/splitlearninghelper.py [hunk 4]
```diff
@@ -0,0 +1,100 @@
+import os
+import tempfile
+import time
+
+import numpy as np
+import torch
```
[the remark this answers is not in the document]
Comment (FrankJonasmoelle) [reply fedn/utils/helpers/plugins/splitlearninghelper.py]: removed this one

**Review by Wrede — APPROVED** [review]
